zeppelin-server/src/main/java/org/apache/zeppelin/realm/jwt/KnoxJwtRealm.java

```diff
@@ -0,0 +1,293 @@
+    return token != null && token instanceof JWTAuthenticationToken;
+  }
+
+
+  @Override
+  protected AuthenticationInfo doGetAuthenticationInfo(AuthenticationToken token) {
+    JWTAuthenticationToken upToken = (JWTAuthenticationToken) token;
+
+    if (validateToken(upToken.getToken())) {
+      try {
+        SimpleAccount account = new SimpleAccount(getName(upToken), upToken.getToken(), getName());
+        account.addRole(mapGroupPrincipals(getName(upToken)));
+        return account;
+      } catch (ParseException e) {
+        LOGGER.error("ParseException in doGetAuthenticationInfo", e);
+      }
+    }
+    return null;
+  }
+
+  private String getName(JWTAuthenticationToken upToken) throws ParseException {
+    SignedJWT signed = SignedJWT.parse(upToken.getToken());
+    String userName = signed.getJWTClaimsSet().getSubject();
+    return userName;
+  }
+
+  private boolean validateToken(String token) {
+    try {
+      SignedJWT signed = SignedJWT.parse(token);
+      return validateSignature(signed);
+    } catch (ParseException ex) {
+      LOGGER.info("ParseException in validateToken", ex);
+      return false;
+    }
+  }
+
+  public static RSAPublicKey parseRSAPublicKey(String pem)
+      throws CertificateException, IOException,
+      ServletException {
+    String PEM_HEADER = "-----BEGIN CERTIFICATE-----\n";
+    String PEM_FOOTER = "\n-----END CERTIFICATE-----";
+    String fullPem = PEM_HEADER + pem + PEM_FOOTER;
+    PublicKey key = null;
+    try {
+      CertificateFactory fact = CertificateFactory.getInstance("X.509");
+      ByteArrayInputStream is = new ByteArrayInputStream(
+          FileUtils.readFileToString(new File(pem)).getBytes("UTF8"));
+      X509Certificate cer = (X509Certificate) fact.generateCertificate(is);
+      key = cer.getPublicKey();
+    } catch (CertificateException ce) {
+      String message = null;
+      if (pem.startsWith(PEM_HEADER)) {
+        message = "CertificateException - be sure not to include PEM header "
+            + "and footer in the PEM configuration element.";
+      } else {
+        message = "CertificateException - PEM may be corrupt";
+      }
+      throw new ServletException(message, ce);
+    } catch (UnsupportedEncodingException uee) {
+      throw new ServletException(uee);
+    } catch (IOException e) {
+      throw new IOException(e);
+    }
+    return (RSAPublicKey) key;
+  }
+
+  protected boolean validateSignature(SignedJWT jwtToken) {
+    boolean valid = false;
+    if (JWSObject.State.SIGNED == jwtToken.getState()) {
+
+      if (jwtToken.getSignature() != null) {
+
+        try {
+          RSAPublicKey publicKey = parseRSAPublicKey(publicKeyPath);
+          JWSVerifier verifier = new RSASSAVerifier(publicKey);
+          if (verifier != null && jwtToken.verify(verifier)) {
+            valid = true;
+
+          } else {
+
+          }
+        } catch (Exception e) {
+          LOGGER.info("Exception in validateSignature nothing to worry", e);
```

Comment:
bad signature is not a concern?

Reply:
This would be most likely problem with the configuration, let me change the text.